modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+    $expiry.parent().show();
+  } else {
+    $expiry.parent().hide();
+    $expiry.val($(e.target).val());
+  }
+  e.target.form.block_timer.value = $expiry.val();
+};
+
+Twinkle.block.callback.update_form = function twinkleblockCallbackUpdateForm(e, data) {
+  var $form = $(e.target.form), $expiry_preset = $form.find('[name=expiry_preset]');
+
+  $expiry_preset.val(data.expiry || 'custom');
+
+  $form.find('[name=expiry]').val(data.expiry);
+  if ($expiry_preset.val() === 'custom') {
+    $form.find('[name=expiry]').parent().show();
+  } else {
+    $form.find('[name=expiry]').parent().hide();
+  }
+
+  // boolean-flipped options, more at [[mw:API:Block]]
+  data.disabletalk = data.disabletalk !== undefined ? data.disabletalk : (data.allowusertalk !== undefined ? data.allowusertalk !== '' : false);
+  data.hardblock = data.hardblock !== undefined ? data.hardblock : (data.anononly !== undefined ? data.anononly !== '' : false);
+
+  $form.find('[name=field_block_options]').find(':checkbox').each(function(i, el) {
+    // don't override original options if useInitialOptions is set
+    if (data.useInitialOptions && data[el.name] === undefined) return;
+    var check = data[el.name] === '' || !!data[el.name];
+    $(el).prop('checked', check);
+  });
+  $form.find('[name=reason]').val(data.reason);
+};
+
+Twinkle.block.callback.change_subcategory = function twinkleblockcallbackChangeSubcategory(e) {
+  var form = e.target.form, prev_block_timer = Twinkle.warn.prev_block_timer;
+  form.main_group.value = 'block';
+  Twinkle.warn.callback.change_subcategory(e);
+  if (prev_block_timer) form.expiry.value = prev_block_timer;
+};
+
+Twinkle.block.callback.evaluate = function twinkleblockCallbackEvaluate(e) {
+  var $form = $(e.target),
+      toBlock = $form.find('[name=actiontype][value=block]').is(':checked'),
+      toWarn = $form.find('[name=actiontype][value=template]').is(':checked');
+
+  if (!toBlock && !toWarn) return alert("Please give Twinkle something to do!");
+
+  if (toBlock) {
+    var blockoptions = $form.find('[name=field_block_options]').serializeObject(),
+        warnoptions = $form.find('[name=field_template_options]').serializeObject();
+    delete blockoptions.expiry_preset; // remove extraneous
+
+    if (!blockoptions.expiry) return alert('Please provide an expiry!');
+    if (blockoptions.reason || (!blockoptions.reason && confirm('No block reason was provided. Continue?'))) {
+      Morebits.simpleWindow.setButtonsEnabled( false );
+      Morebits.status.init( e.target );
+      formCleared = true;
+      blockoptions.action = 'block';
+      blockoptions.user = relevantUser;
+
+      // boolean-flipped options
+      blockoptions.anononly = blockoptions.hardblock ? undefined : true;
+      blockoptions.allowusertalk = blockoptions.disabletalk ? undefined : true;
+
+      // temporary fix for bug with block API, see [[phab:T68646]]
+      if (blockoptions.expiry === 'infinity') blockoptions.expiry = 'infinite';
```

Comment:
From what I've been told it will be deployed with 1.25wmf22, due to released on all wikis 25 March https://www.mediawiki.org/wiki/MediaWiki_1.25/Roadmap